Generate some PropertyDeclaration methods by hand 🐉🐲

nox opened the pull request:
We rely on https://github.com/rust-lang/rfcs/pull/2195 to make some enums share the same discriminants by definition.

This fixes https://bugzilla.mozilla.org/show_bug.cgi?id=1428285.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19956)
<!-- Reviewable:end -->

nox committed:
Use less type aliases from longhands in PropertyDeclaration

This will help us reuse code between different variants in a later patch,
to implement Clone by hand to optimise it.

nox committed:
Use LonghandId::is_animatable in PropertyDeclaration::is_animatable

nox committed:
Make all PropertyDeclaration variants have only one field

nox committed:
Implement Clone for PropertyDeclaration by hand 🐉🐲

This makes the enum `#[repr(u16)]` and generate code that shares the same
Clone::clone call for all variants sharing the same field type.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/data.py, components/style/properties/helpers.mako.rs, components/style/properties/declaration_block.rs, components/style/properties/helpers/animated_properties.mako.rs, components/style/properties/properties.mako.rs
 * @canaltinova: components/style/properties/data.py, components/style/properties/helpers.mako.rs, components/style/properties/declaration_block.rs, components/style/properties/helpers/animated_properties.mako.rs, components/style/properties/properties.mako.rs
 * @emilio: components/style/properties/data.py, components/style/properties/helpers.mako.rs, components/style/properties/declaration_block.rs, components/style/properties/helpers/animated_properties.mako.rs, components/style/properties/properties.mako.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

nox commented:
https://treeherder.mozilla.org/#/jobs?repo=try&revision=0cccd012152be2f205e31bb00254900e85215152

SimonSapin commented on components/style/properties/properties.mako.rs:
```diff
@@ -1438,34 +1441,126 @@ impl PropertyId {
 #[cfg_attr(feature = "gecko", derive(MallocSizeOf))]
-#[derive(Clone, PartialEq)]
+#[derive(PartialEq)]
+#[repr(u8)]
```
Doesn’t this have more than 256 variants?

nox replied:
Yes. :D Just realised with the geckotry failing to build.

nox commented:
I made the enum be `#[repr(u16)]`.

https://treeherder.mozilla.org/#/jobs?repo=try&revision=680ae344ebda08a71b94be64049039d2328abd33

nox committed:
Make LonghandId just be the discriminants of PropertyDeclaration 🐉🐲

This relies on the fact that `#[repr(u16)]` on enums ensure the discriminant will
be the very first field of the enum value. This has always been the case, but it
has been formally defined only since https://github.com/rust-lang/rfcs/pull/2195.

https://bugzilla.mozilla.org/show_bug.cgi?id=1428285

nox committed:
Merge common arms in PropertyDeclaration::to_css

nox committed:
Implement PartialEq for PropertyDeclaration by hand 🐉🐲

We first check that the discriminants are equal, then we pattern match only
on `*self` and use `PropertyDeclarationVariantRepr` to look into `other` directly.

nox commented:
```
[nox:~/src/gecko] 2s 130
⌀ size XUL.old
__TEXT	__DATA	__OBJC	others	dec	hex
88387584	5271552	0	65503232	159162368	97ca000

[nox:~/src/gecko]
⌀ ../bloaty/build/bloaty -d symbols -n 100000000 XUL.old | grep -w PropertyDeclaration
   0.0%  45.1Ki style::properties::PropertyDeclaration::to_css::h70aa9127e631f236                 45.1Ki   0.0%
   0.0%  32.8Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialEq$GT$:  32.8Ki   0.0%
   0.0%  32.6Ki __ZN79_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialE  32.6Ki   0.0%
   0.0%  29.7Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  29.7Ki   0.0%
   0.0%  29.3Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  29.3Ki   0.0%
   0.0%  27.9Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  27.9Ki   0.0%
   0.0%  27.6Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$GT$::c  27.6Ki   0.0%
   0.0%  15.7Ki __ZN87_$LT$style..properties..PropertyDeclaration$u20$as$u20$malloc_size_of..Mal  15.7Ki   0.0%
   0.0%  9.78Ki style::properties::PropertyDeclaration::parse_into::h35ee1deaf54c46b6             9.78Ki   0.0%
   0.0%  4.17Ki style::properties::PropertyDeclaration::id::h9f691ba9ad1fcbbc                     4.17Ki   0.0%
   0.0%     320 style::properties::PropertyDeclaration::get_system::h4a3d3afe5a217d86                320   0.0%
   0.0%      32 style::properties::PropertyDeclaration::get_css_wide_keyword::he3ca52066fa1abf7       32   0.0%

[nox:~/src/gecko] 2s
⌀ size XUL.new
__TEXT	__DATA	__OBJC	others	dec	hex
88190976	5271552	0	65482752	158945280	9795000

[nox:~/src/gecko]
⌀ ../bloaty/build/bloaty -d symbols -n 100000000 XUL.new | grep -w PropertyDeclaration
   0.0%  45.2Ki style::properties::PropertyDeclaration::to_css::hc9cb9df128aea03f                 45.2Ki   0.0%
   0.0%  22.8Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialEq$GT$:  22.8Ki   0.0%
   0.0%  22.3Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$GT$::c  22.3Ki   0.0%
   0.0%  15.6Ki __ZN87_$LT$style..properties..PropertyDeclaration$u20$as$u20$malloc_size_of..Mal  15.6Ki   0.0%
   0.0%  9.89Ki style::properties::PropertyDeclaration::parse_into::ha8f87489fe4f89f8             9.89Ki   0.0%
   0.0%     320 style::properties::PropertyDeclaration::get_system::h07f9efc836477132                320   0.0%
   0.0%      96 style::properties::PropertyDeclaration::id::hd928f19c20f0e1bc                         96   0.0%
   0.0%      32 style::properties::PropertyDeclaration::get_css_wide_keyword::hfd74e2964361369d       32   0.0%
```

It looks good but I'm not sure I should trust those results.

Cc @bholley @emilio 

Also, I've yet to make a geckotry with the latest commits I added.

https://treeherder.mozilla.org/#/jobs?repo=try&revision=d0b009ecd44b0a7327c507b37a9b188e2a5b7e7b

nox commented:
Silly me had not put `#[inline]` on the `Clone` and `PartialEq` methods, that explains why there was only one `Clone` in `XUL.new`:

```
⌀ ../bloaty/build/bloaty -d symbols -n 1000000000 XUL.new | grep -w PropertyDeclaration
   0.0%  45.2Ki style::properties::PropertyDeclaration::to_css::hc9cb9df128aea03f                 45.2Ki   0.0%
   0.0%  23.0Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialEq$GT$:  23.0Ki   0.0%
   0.0%  22.7Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.7Ki   0.0%
   0.0%  22.7Ki __ZN79_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialE  22.7Ki   0.0%
   0.0%  22.3Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.3Ki   0.0%
   0.0%  22.2Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.2Ki   0.0%
   0.0%  21.8Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$GT$::c  21.8Ki   0.0%
   0.0%  15.6Ki __ZN87_$LT$style..properties..PropertyDeclaration$u20$as$u20$malloc_size_of..Mal  15.6Ki   0.0%
   0.0%  9.89Ki style::properties::PropertyDeclaration::parse_into::ha8f87489fe4f89f8             9.89Ki   0.0%
   0.0%     320 style::properties::PropertyDeclaration::get_system::h07f9efc836477132                320   0.0%
   0.0%      96 style::properties::PropertyDeclaration::id::hd928f19c20f0e1bc                         96   0.0%
   0.0%      32 style::properties::PropertyDeclaration::get_css_wide_keyword::hfd74e2964361369d       32   0.0%
```

nox commented:
Everything is orange. Will investigate tomorrow. Probably some dumb error somewhere.

nox committed:
Generate PropertyDeclaration just after processing all the Mako files

nox commented:
Latest results with bloaty, and the missing `#[inline]` on `clone` and `eq`:

```
[nox:~/src/gecko]
⌀ size XUL.old
__TEXT	__DATA	__OBJC	others	dec	hex
88387584	5271552	0	65503232	159162368	97ca000

[nox:~/src/gecko]
⌀ ../bloaty/build/bloaty -d symbols -n 1000000000 XUL.old | grep -w PropertyDeclaration
   0.0%  45.1Ki style::properties::PropertyDeclaration::to_css::h70aa9127e631f236                 45.1Ki   0.0%
   0.0%  32.8Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialEq$GT$:  32.8Ki   0.0%
   0.0%  32.6Ki __ZN79_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialE  32.6Ki   0.0%
   0.0%  29.7Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  29.7Ki   0.0%
   0.0%  29.3Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  29.3Ki   0.0%
   0.0%  27.9Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  27.9Ki   0.0%
   0.0%  27.6Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$GT$::c  27.6Ki   0.0%
   0.0%  15.7Ki __ZN87_$LT$style..properties..PropertyDeclaration$u20$as$u20$malloc_size_of..Mal  15.7Ki   0.0%
   0.0%  9.78Ki style::properties::PropertyDeclaration::parse_into::h35ee1deaf54c46b6             9.78Ki   0.0%
   0.0%  4.17Ki style::properties::PropertyDeclaration::id::h9f691ba9ad1fcbbc                     4.17Ki   0.0%
   0.0%     320 style::properties::PropertyDeclaration::get_system::h4a3d3afe5a217d86                320   0.0%
   0.0%      32 style::properties::PropertyDeclaration::get_css_wide_keyword::he3ca52066fa1abf7       32   0.0%

[nox:~/src/gecko] 3s
⌀ size XUL.new
__TEXT	__DATA	__OBJC	others	dec	hex
88317952	5271552	0	65495040	159084544	97b7000

[nox:~/src/gecko]
⌀ ../bloaty/build/bloaty -d symbols -n 1000000000 XUL.new | grep -w PropertyDeclaration
   0.0%  45.2Ki style::properties::PropertyDeclaration::to_css::he42a36e94b3deab2                 45.2Ki   0.0%
   0.0%  23.0Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialEq$GT$:  23.0Ki   0.0%
   0.0%  22.7Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.7Ki   0.0%
   0.0%  22.7Ki __ZN79_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialE  22.7Ki   0.0%
   0.0%  22.3Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.3Ki   0.0%
   0.0%  22.2Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.2Ki   0.0%
   0.0%  21.8Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$GT$::c  21.8Ki   0.0%
   0.0%  15.6Ki __ZN87_$LT$style..properties..PropertyDeclaration$u20$as$u20$malloc_size_of..Mal  15.6Ki   0.0%
   0.0%  9.89Ki style::properties::PropertyDeclaration::parse_into::ha8f87489fe4f89f8             9.89Ki   0.0%
   0.0%     320 style::properties::PropertyDeclaration::get_system::h07f9efc836477132                320   0.0%
   0.0%      96 style::properties::PropertyDeclaration::id::hd928f19c20f0e1bc                         96   0.0%
   0.0%      32 style::properties::PropertyDeclaration::get_css_wide_keyword::hfd74e2964361369d       32   0.0%
```

Geckotry seems green: https://treeherder.mozilla.org/#/jobs?repo=try&revision=2cb22ae5e6be97fcc968dc965827234d54f26b47&selectedJob=160528329

nox committed:
Implement MallocSizeOf for PropertyDeclaration by hand

We merge arms with identical field types.

emilio commented on components/style/properties/helpers.mako.rs:
```diff
@@ -903,6 +903,7 @@
     <%call expr="longhand(name,
                           predefined_type=length_type,
                           logical=logical,
+                          is_size_type=True,
```
Let's rename this to something that screams out more, like `is_gecko_size_type_hack` or something.

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -1441,28 +1442,94 @@ impl PropertyId {
-    %>
-    /// ${property.name}
-    ${property.camel_case}(${ty}),
+        variants.append({
```
Can you document why do we need this complexity here, please?

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -1623,23 +1646,9 @@ impl PropertyDeclaration {
-        };
-        PropertyDeclarationId::Longhand(longhand_id)
+        PropertyDeclarationId::Longhand(
+            unsafe { *(self as *const _ as *const LonghandId) },
```
Can we debug-assert this somehow at least, to ensure nobody messes up? I guess if they mess up it'd be completely orange but...

Also, please add a comment about why this is ok here so people don't scream :).

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -1572,6 +1599,28 @@ pub struct VariableDeclaration {
+        match self.value.from_shorthand {
+            // Normally, we shouldn't be printing variables here if they came from
+            // shorthands. But we should allow properties that came from shorthand
+            // aliases. That also matches with the Gecko behavior.
```
Can you mention here that this is just a hack for `-moz-transform`, which is really a shorthand for who knows what reason?

Also we should try to unship that, so feel free to also add a `FIXME(emilio)` over there so I eventually do it.

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -314,6 +313,32 @@ impl Clone for PropertyDeclaration {
+                ${" | ".join("{}(ref this)".format(v) for v in variants)} => {
+                    let other_repr =
+                        &*(other as *const _ as *const PropertyDeclarationVariantRepr<${ty}>);
+                    *this == other_repr.value
```
Wonderfully evil :(

emilio commented on components/style/properties/properties.mako.rs:
```diff
@@ -267,7 +268,6 @@ pub mod animated_properties {
 /// Servo's representation for a property declaration.
-#[cfg_attr(feature = "gecko", derive(MallocSizeOf))]
 #[repr(u16)]
 pub enum PropertyDeclaration {
```
Can you add a general overview of why don't we derive stuff here and similar?

emilio reviewed (approved):
r=me, with the comments and a different name for `is_size_type` (feel free to rename it on a followup commit if it's easier).

emilio commented:
If @SimonSapin sanity-checked this too it'd be nice, since he had concerns about doing the same kind of optimization in https://bugzilla.mozilla.org/show_bug.cgi?id=1428285#c2, it seems.

SimonSapin commented:
My concern in that comment was about transmuting the return type of `std::mem::discriminant` to u64. This function was very deliberately designed to not return a plain integer, but instead return an opaque type that makes few guarantees.

This PR however relies on memory layout guarantees documented by https://github.com/rust-lang/rfcs/pull/2195 when `#[repr(u16)]` is used on an enum type.

@nox Please add a link to https://github.com/rust-lang/rfcs/pull/2195 and a short explanation to the PR message.

nox commented:
I added a comment in the commit message.

@bors-servo r=emilio,SimonSapin

bors-servo commented:
:pushpin: Commit fd31712 has been approved by `emilio,SimonSapin`

<!-- @bors-servo r=emilio,SimonSapin fd31712a56da89f013d13cd657ffc9d32bc01c6c -->

bors-servo commented:
:hourglass: Testing commit fd31712a56da89f013d13cd657ffc9d32bc01c6c with merge 5304e8f288e8cd26f35b4e1dc60de5721a95e497...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7812)

nox commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 1f0a126 has been approved by `emilio`

<!-- @bors-servo r=emilio 1f0a1267bb93d7fbf58b1dbd43048e5bdd7be181 -->

bors-servo commented:
:hourglass: Testing commit 1f0a1267bb93d7fbf58b1dbd43048e5bdd7be181 with merge 9faf0cdce50379dfb8125d7a9c913babd56382e2...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10637), [arm32](http://build.servo.org/builders/arm32/builds/10878), [arm64](http://build.servo.org/builders/arm64/builds/10907), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10936), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7801), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7813), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11006), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3404), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3427), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7669), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7738), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3501), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3445), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6317)
Approved by: emilio
Pushing 9faf0cdce50379dfb8125d7a9c913babd56382e2 to master...

ionutgoldan commented:
These changes landed some fine performance improvements! Thank you!

== Change summary for alert #11504 (as of Tue, 06 Feb 2018 12:06:36 GMT) ==

Improvements:

  5%  Stylo Servo_DeclarationBlock_SetPropertyById_WithInitialSpace_Bench osx-10-10 opt      549,031.92 -> 519,840.33
  2%  Stylo Servo_DeclarationBlock_GetPropertyById_Bench windows7-32 opt                     231,626.04 -> 226,308.08

For up to date results, see: https://treeherder.mozilla.org/perf.html#/alerts?id=11504

bholley commented:
There are some more wins in #19966 that might also be attributable to this.

Awesome job @nox!
